src/plugins/intel_gpu/src/plugin/infer_request.cpp

```diff
@@ -855,7 +855,35 @@ void InferRequest::prepare_input(const cldnn::primitive_id& inputName, Blob::Ptr
     };
 
     auto _nw_ptr = m_graph->GetNetwork();
-    if (input_layout.is_dynamic()) {
+    if (node_input_layout.is_dynamic()) {
+        // Reuse previous output usm buffer as current input, if the original user memories for them are same
+        // (i.e., an external loop)
+        for (auto out_mem : _outputs) {
+            auto out_ptr = out_mem.second->cbuffer().as<void*>();
+            auto in_ptr = inputBlob->cbuffer().as<void*>();
+            if (out_ptr == in_ptr) {
+                const cldnn::primitive_id internalInputName = "parameter:" + inputName;
+                auto outputID = outputsMap.empty() ? m_graph->MapOutputName(out_mem.first) : outputsMap.at(out_mem.first);
+                auto inDesc = inputBlob->getTensorDesc();
+                auto in_format = FormatFromLayout(inDesc.getLayout());
+                auto in_dt = DataTypeFromPrecision(inDesc.getPrecision());
+                auto in_layout = cldnn::layout{ov::PartialShape(inDesc.getDims()), in_dt, in_format};
+                auto prev_output_mem = internal_outputs.at(outputID).get_memory(false);
+                auto prev_output_layout = internal_outputs.at(outputID).get_layout();
+                GPU_DEBUG_TRACE_DETAIL << "Use previous output memory of " << outputID
+                                       << " as current input memory for " << internalInputName << std::endl;
+                if (prev_output_layout == in_layout) {
+                    if (prev_output_mem->get_layout() != prev_output_layout) {
+                        prev_output_mem =
+                            m_graph->get_engine().reinterpret_buffer(*prev_output_mem, prev_output_layout);
+                    }
+                    _nw_ptr->set_output_buffer_used_for_next_input(outputID, internalInputName);
```

Comment:
That implementation looks quite easy, but honestly, I'd prefer to have buffer swap within infer request only (_device_output and _device_input swap), so the reset of the flow will be exactly the same as for other dynamic cases. That would be much more easy to maintain as swapping logic will be isolated in request.

Likely we'll need an ability to set output memory from request to network for dynamic models as a prerequisite

Reply:
My initial impl was only swap memory, but there were multiple issues needed to be hanlded internally:
-  if the out mem is to be reused as input : it should not be limited to usm_host & should not allow its memory to be reused
- Also if the output is to be used as input of the same node (e.g., concat), it should allocate new memory for the output 

So the current impl is made.. but basically I also agree that we need to seek for better solution. 

>>Likely we'll need an ability to set output memory from request to network for dynamic models as a prerequisite

The output shape is decided at runtime, but how can we  set output memory from infer request ?